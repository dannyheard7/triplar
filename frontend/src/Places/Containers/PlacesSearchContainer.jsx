import React from "react";
import api from "Places/utils/places.api.js";
import PlaceListContainer from "Places/Containers/PlaceListContainer";

import "Places/styles/places.css";
import {getPopularPlaces, getTopLevelCategories} from "../utils/actions";
import { connect } from "react-redux";

export class PlacesSearchContainer extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            searchPlaces: [],
            selectedCategory: "",
            selectedSubCategory: "",
            subCategories: [],
            searchValue: "",
        };

        this.searchPlaces = this.searchPlaces.bind(this);
        this.getPopularPlaces = this.getPopularPlaces.bind(this);
        this.onSearchChange = this.onSearchChange.bind(this);
        this.onCategoryChange = this.onCategoryChange.bind(this);
        this.onSubCategoryChange = this.onSubCategoryChange.bind(this);
    }

    componentDidMount() {
        this.getPopularPlaces(this.state.selectedCategory);
        this.props.dispatch(getTopLevelCategories());
    }

    getPopularPlaces(category) {
        const city = this.props.city;
        this.props.dispatch(getPopularPlaces(city.location.lat, city.location.lng, category));
    }

    searchPlaces(category) {
        // Searches are not cached because they are not used multiple times!
        const city = this.props.city;
        const search = this.state.searchValue;

        api.searchPlacesByName(city.location.lat, city.location.lng, search, category).then(response => {
            this.setState({searchPlaces: response.data.data.places})
        });
    };

    getSubCategories() {
        api.getSubCategories(this.state.selectedCategory, this.props.city.country.code).then(response => {
            this.setState({subCategories: response.data.data.subCategories})
        });
    }

    updatePlaces() {
        let category = this.state.selectedSubCategory ? this.state.selectedSubCategory : this.state.selectedCategory;

        if (this.state.searchValue.length >= 3) {
            this.searchPlaces(category);
        } else {
             this.getPopularPlaces(category);
        }
    }

    onSearchChange(event) {
        this.setState({searchValue:  event.target.value}, this.updatePlaces);
    }

    onCategoryChange(event) {
        this.setState({selectedCategory: event.target.value, selectedSubCategory: "", subCategories: []},
            () => {this.updatePlaces(); this.getSubCategories();});
    }

    onSubCategoryChange(event) {
        this.setState({selectedSubCategory: event.target.value},  this.updatePlaces);
    }

    render() {
        const city = this.props.city;
        let category = this.state.selectedSubCategory ? this.state.selectedSubCategory : this.state.selectedCategory;

        let places = [];
        if (this.state.searchValue.length >= 3) {
            places = this.state.searchPlaces;
        } else {
            let popularPlaces = this.props.popularPlaces.find(x => x.lat === city.location.lat &&
                    x.lng === city.location.lng && x.category === category);
            if(popularPlaces) {
                places = popularPlaces.places.map(x => this.props.places.find(p => p.id === x));
            }
        }

        return (
            <div className="places-search-container">
                <h3>Popular Places in {city.name}</h3>
                <input type="text" value={this.state.searchValue} onChange={this.onSearchChange}
                       placeholder="Search Places (min 3 characters)"/>
                <div>
                    <select onChange={this.onCategoryChange} defaultValue={this.state.selectedCategory} id="category-select">
                        <option value="">Any Category</option>
                        {this.props.categories.map(this.categoryAsOption)}
                    </select>
                    {this.state.selectedCategory.length > 0 &&
                        <select onChange={this.onSubCategoryChange} defaultValue={this.state.selectedSubCategory}
                                id="subcategory-select">
                            <option value="">Any Subcategory</option>
                            {this.state.subCategories && this.state.subCategories.map(this.categoryAsOption)}
                        </select>
                    }
                </div>
                <PlaceListContainer places={places} path={this.props.path} />
            </div>
        );
    }

    categoryAsOption(category) {
        return <option value={category.alias} key={category.alias}>{category.title}</option>
    }
}

PlacesSearchContainer.defaultProps = {
    popularPlaces: [],
    categories: []
};

const mapStateToProps = (state) => ({
    popularPlaces: state.places.popularPlaces,
    places: state.places.places,
    categories: state.places.topLevelCategories.categories
});

export default connect(mapStateToProps)(PlacesSearchContainer)