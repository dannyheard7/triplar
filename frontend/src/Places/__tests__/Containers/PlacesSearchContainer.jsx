import React from "react";
import {shallow} from "enzyme";

import ShallowRenderer from 'react-test-renderer/shallow';
import PlacesSearchContainer from "Places/Containers/PlacesSearchContainer";

const renderer = new ShallowRenderer();

const faker = require('faker');
jest.mock('Places/utils/places.api.js');

const api = require('Places/utils/places.api.js');

describe('<PlacesSearchContainer />', () => {
    const event = {
        preventDefault: () => {},
        target: {value: "", name: ""},
    };

    const city = {name: faker.address.city(), country: {name: faker.address.country(), code: faker.address.countryCode},
        location: {lat: faker.address.latitude, lng: faker.address.longitude}};
    const places = [{'id': faker.random.number(), 'name': faker.random.word(), 'imageUrl': faker.internet.url()}];

    test('renders correctly', () => {
        const city = {name: 'Bristol', country: {name: 'United Kingdom'}};
        const result = renderer.render(<PlacesSearchContainer city={city} />);
        expect(result).toMatchSnapshot();
    });

    test('getPopularPlaces is called on mount', () => {
        const getPopularPlacesStub = jest.spyOn(PlacesSearchContainer.prototype, 'getPopularPlaces');
        const container = shallow(<PlacesSearchContainer city={city} />);
        expect(getPopularPlacesStub).toBeCalled();

        getPopularPlacesStub.mockRestore();
    });

    test('search onChange  calls searchPlaces if length >= 3', () => {
        const stub = jest.spyOn(PlacesSearchContainer.prototype, 'searchPlaces');
        const container = shallow(<PlacesSearchContainer city={city} />);

        event.target.value = "abc";
        container.find("input").prop('onChange')(event);

        expect(container.state('searchValue')).toEqual(event.target.value);
        expect(stub).toBeCalled();

        stub.mockRestore();
    });

    test('search onChange does not call searchPlaces if length < 3', () => {
        const stub = jest.spyOn(PlacesSearchContainer.prototype, 'searchPlaces');
        const container = shallow(<PlacesSearchContainer city={city} />);

        container.setState({popularPlacesCache: {"": places}});
        event.target.value = "ab";
        container.find("input").prop('onChange')(event);

        expect(stub).not.toBeCalled();
        expect(container.state('places')).toEqual(places);

        stub.mockRestore();
    });

    test('updatePlaces calls getPopularPlaces with subCategory not main category if one is set', () => {
        const getPopularPlacesStub = jest.spyOn(PlacesSearchContainer.prototype, 'getPopularPlaces');
        const container = shallow(<PlacesSearchContainer city={city} />);
        getPopularPlacesStub.mockClear(); // It is called on mount

        let subCategory = faker.lorem.word();
        container.setState({selectedSubCategory: subCategory});
        container.instance().updatePlaces();

        expect(getPopularPlacesStub).toBeCalledWith(subCategory);
        getPopularPlacesStub.mockRestore();
    });

    test('updatePlaces calls getPopularPlaces with selectedCategory if searchValue length is < 3', () => {
        const searchPlacesStub = jest.spyOn(PlacesSearchContainer.prototype, 'searchPlaces');
        const getPopularPlacesStub = jest.spyOn(PlacesSearchContainer.prototype, 'getPopularPlaces');
        const container = shallow(<PlacesSearchContainer city={city} />);
        getPopularPlacesStub.mockClear(); // It is called on mount

        container.instance().updatePlaces();

        expect(getPopularPlacesStub).toBeCalled();
        expect(searchPlacesStub).not.toBeCalled();

        searchPlacesStub.mockRestore();
        getPopularPlacesStub.mockRestore();
    });

    test('updatePlaces calls searchPlaces if searchValue length is > 3', () => {
        const searchPlacesStub = jest.spyOn(PlacesSearchContainer.prototype, 'searchPlaces');
        const getPopularPlacesStub = jest.spyOn(PlacesSearchContainer.prototype, 'getPopularPlaces');
        const container = shallow(<PlacesSearchContainer city={city} />);
        getPopularPlacesStub.mockClear(); // It is called on mount

        container.setState({searchValue: 'abcd'});
        container.instance().updatePlaces();

        expect(searchPlacesStub).toBeCalled();
        expect(getPopularPlacesStub).not.toBeCalled();

        searchPlacesStub.mockRestore();
        getPopularPlacesStub.mockRestore();
    });

    test('getPopularPlaces caches returned value from API', async () => {
        const stub = jest.spyOn(api.default, "getPopularPlaces");
        stub.mockReturnValueOnce(Promise.resolve({status: 200, data: {data: {popularPlaces: places}}}));

        const container = shallow(<PlacesSearchContainer city={city} />);
        await Promise.resolve();

        expect(stub).toBeCalledWith(city.location.lat, city.location.lng, "");
        expect(container.state('popularPlacesCache')).toEqual({"": places});

        stub.mockRestore();
    });

    test('getPopularPlaces uses cached value if it exists', async () => {
        const container = shallow(<PlacesSearchContainer city={city} />);
        await Promise.resolve(); // Initial call to the api onMount

        const spy = jest.spyOn(api.default, "getPopularPlaces");
        container.setState({popularPlacesCache: {"": places}});
        container.instance().getPopularPlaces("");

        expect(spy).not.toBeCalled();
        expect(container.state('places')).toEqual(places);
    });


    test('on category change, corresponding subcategories are updated from api', async () => {
        const container = shallow(<PlacesSearchContainer city={city} />);
        await Promise.resolve(); // Initial call to the api onMount

        event.target.value = "ab";

        const spy = jest.spyOn(api.default, "getSubCategories");
        let subCategories = [{name: faker.lorem.word(), alias: faker.lorem.word()}];
        spy.mockReturnValueOnce(Promise.resolve({status: 200, data: {data: {subCategories}}}));

        container.instance().onCategoryChange(event);

        await Promise.resolve();
        expect(spy).toBeCalledWith(event.target.value, city.country.code);
        expect(container.state('subCategories')).toEqual(subCategories);
    });
});