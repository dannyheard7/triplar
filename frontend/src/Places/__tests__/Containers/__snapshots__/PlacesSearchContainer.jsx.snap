// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PlacesSearchContainer /> renders correctly 1`] = `
<div
  className="places-search-container"
>
  <h3>
    Popular Places in 
    Bristol
  </h3>
  <input
    onChange={[Function]}
    placeholder="Search Places (min 3 characters)"
    type="text"
    value=""
  />
  <div>
    <select
      defaultValue=""
      id="category-select"
      onChange={[Function]}
    >
      <option
        value=""
      >
        Any Category
      </option>
    </select>
  </div>
  <DroppablePlaceListContainer
    droppableId="placeSearchContainer"
    onDrop={[Function]}
    places={Array []}
  />
</div>
`;
